src/3d/Options.fst

```diff
@@ -9,33 +9,80 @@ let input_file : ref (list string) = alloc []
+   (noshort, "batch", ZeroArgs (fun _ -> batch := true), "Verify the generated F* code and extract C code");
+   (noshort, "no_batch", ZeroArgs (fun _ -> batch := false), "Do not verify the generated F* code or extract C code");
+   (noshort, "clang_format", ZeroArgs (fun _ -> batch := true; clang_format := true), "Call clang-format on extracted .c/.h files (--batch only)");
+   (noshort, "no_clang_format", ZeroArgs (fun _ -> clang_format := false), "Do not call clang-format on extracted .c/.h files");
+   (noshort, "clang_format_executable", OneArg ((fun cmd -> batch := true; clang_format := true; clang_format_executable := cmd), "clang-format full path"), "Provide path to clang-format if not reachable through PATH");
+   (noshort, "cleanup", ZeroArgs (fun _ -> cleanup := true), "Remove *.fst*, *.krml and kremlin-args.rsp (--batch only)");
+   (noshort, "no_cleanup", ZeroArgs (fun _ -> cleanup := false), "Do not remove *.fst*, *.krml or kremlin-args.rsp");
+   (noshort, "version", ZeroArgs (fun _ -> FStar.IO.print_string (Printf.sprintf "EverParse/3d %s\nCopyright 2018, 2019, 2020 Microsoft Corporation\n" Version.everparse_version); exit 0), "Show this version of EverParse");
    ]
 
+let options : ref _ = alloc options0
+
 let display_usage () : ML unit =
   FStar.IO.print_string "3d [options] input_file\n";
   List.iter (fun (_, m, _, desc) ->
     FStar.IO.print_string (Printf.sprintf "    --%s %s\n" m desc))
-    options
+    !options
+      ;
+  if !batch then begin
+    FStar.IO.print_string "--batch is currently toggled.\n";
+    if !clang_format then begin
+      FStar.IO.print_string "--clang_format is currently toggled.\n"
+    end;
+    if !cleanup then begin
+      FStar.IO.print_string "--cleanup is currently toggled.\n"
+    end
+  end
+
+let _ =
+  options := ('h', "help", FStar.Getopt.ZeroArgs (fun _ -> display_usage (); exit 0), "Show this help message") :: !options
 
 let parse_cmd_line () : ML (list string) =
   let open FStar.Getopt in
-  let res = FStar.Getopt.parse_cmdline options (fun file -> input_file := file :: !input_file) in
+  let res = FStar.Getopt.parse_cmdline !options (fun file -> input_file := file :: !input_file) in
   match res with
   | Success -> !input_file
   | Help -> display_usage(); exit 0
   | Error s -> FStar.IO.print_string s; exit 1
   | _ -> exit 2
 
+let rec list_last #a (l: list a { Cons? l }) : Tot a =
+  match l with
+  | [x] -> x
+  | _ :: l' -> list_last l'
+
+let basename (fn: string) : Tot string =
+  let fn2 =
+    match String.split ['\\'] fn with
+    | [] -> fn
+    | l -> list_last l
+  in
+  match String.split ['/'] fn2 with
+  | [] -> fn2
+  | l2 -> list_last l2
+
 let split_3d_file_name fn =
-  match String.split ['.'] fn with
+  match String.split ['.'] (basename fn) with
   | [name;extension] ->
     // FStar.IO.print_string
     //   (Printf.sprintf "filename = %s; name=%s; extension=%s"

```

Comment:
it is intended to fail anyway because the basename of the file, which will be reused as a component of a C identifier, should not have `.`
Maybe we should have an explicit validation step on what we expect to be a valid module name.

Reply:
early validation sounds good rather than late-failure